pkg/backend/spinnaker/backend.go

```diff
@@ -856,3 +864,186 @@ func (s *SpinClient) SavePipeline(pipelineJSON string) (*http.Response, error) {
 
 	return res, nil
 }
+
+// A DFS implementation that runs through the pipeline/stages tree.
+// Finds child pipelines and deletes them.
+// Each iteration of the stages loop will look for "pipeline" key in each element
+// If it finds one and it is another pipeline it will start another iteration on that pipeline's stages
+// Once a pipeline with no "pipeline" stages is met - it is deleted
+// The parent loop continues until all its stages of type pipeline are deleted.
+// Once all loops are closed the most top level pipeline gets deleted.
+func (s *SpinClient) getNestedPipelinesNames(stages interface{}, pipeline map[string]interface{}) ([]string, error) {
+	pipelineNames := []string{}
+	for _, stage := range stages.([]interface{}) {
+		stage := stage.(map[string]interface{})
+		stagePipelineField, exists := stage["pipeline"]
+		if !exists {
+			continue
+		}
+
+		switch stagePipelineField.(type) {
+		case string:
+			continue
+		}
+
+		childPipeline := stage["pipeline"].(map[string]interface{})
+
+		if err := s.isValidPipeline(childPipeline); err != nil {
+			return pipelineNames, err
+		}
+
+		parentPipelineApplication := []string{pipeline["application"].(string)}
+
+		if err := isValidPipelineApplication(childPipeline, parentPipelineApplication); err != nil {
+			return pipelineNames, err
+		}
+
+		childPipelineStages, exists := childPipeline["stages"]
+		if exists {
+			hasChildPipelines, err := hasValidChildPipelineStages(childPipelineStages.([]interface{}), parentPipelineApplication)
+			if err != nil {
+				return pipelineNames, err
+			}
+
+			// If any of stages is of type pipeline get those pipelines names recursively
+			if hasChildPipelines {
+				nestedPipelineNames, err := s.getNestedPipelinesNames(childPipelineStages, childPipeline)
+				if err != nil {
+					return pipelineNames, err
+				}
+				pipelineNames = append(pipelineNames, nestedPipelineNames...)
+			}
+		}
+
+		// After we return from recursion we add "this layer" child pipeline
+		childPipelineName := childPipeline["name"].(string)
+		pipelineNames = append(pipelineNames, childPipelineName)
+	}
+
+	return pipelineNames, nil
```

Comment:
Duplicate names need to be removed - otherwise, it will fail on deleting a pipeline that doesn't exist.

Reply:
Might be better to de-duplicate on line 984 / in `DeletePipeline`.

Reply:
after the fix:
```shell
Application: testapp
Pipelines to delete: [yolo-pipeline childPipelineLevel3 childPipelineLevel2 childPipelineLevel1 deletetest]
DELETED: testapp - childPipelineLevel1
DELETED: testapp - childPipelineLevel3
DELETED: testapp - yolo-pipeline
DELETED: testapp - childPipelineLevel2
DELETED: testapp - deletetest
```